File: test-user-input-runner.py
#!/usr/bin/env python3

"""
Test User Input Runner - Python Version
This script demonstrates how to programmatically submit test cases to the security testing platform
"""

import json
import time
import argparse
import requests
from typing import Dict, List, Any
import os
from datetime import datetime

# Configuration
API_BASE_URL = os.getenv('API_URL', 'http://localhost:3000/api')
DELAY_BETWEEN_TESTS = 2  # seconds

class SecurityTestRunner:
    def __init__(self, api_url: str = API_BASE_URL):
        self.api_url = api_url
        self.session = requests.Session()
        self.results = {
            'successful': [],
            'failed': [],
            'total': 0,
            'start_time': datetime.now()
        }
    
    def create_workflow_payload(self, test_input: Dict[str, Any]) -> Dict[str, Any]:
        """Create the API payload from test input"""
        payload = {
            'target': test_input['target'],
            'scope': test_input['scope'],
            'description': test_input['description'],
            'template': f"security-{test_input['testType']}"
        }
        
        # Add authentication if provided
        if test_input.get('username') and test_input.get('password'):
            payload['auth'] = {
                'username': test_input['username'],
                'password': test_input['password']
            }
        
        return payload
    
    def run_test(self, scenario: Dict[str, Any]) -> Dict[str, Any]:
        """Run a single test scenario"""
        print(f"\n🔍 Running: {scenario['scenario']}")
        print(f"📝 Description: {scenario['description']}")
        
        try:
            payload = self.create_workflow_payload(scenario['input'])
            print(f"📤 Payload: {json.dumps(payload, indent=2)}")
            
            response = self.session.post(
                f"{self.api_url}/run-soc2-workflow",
                json=payload,
                timeout=30
            )
            response.raise_for_status()
            
            data = response.json()
            print(f"✅ Success! Workflow ID: {data.get('workflowId')}")
            print(f"📊 Status: {data.get('status')}")
            
            return {
                'success': True,
                'workflowId': data.get('workflowId'),
                'scenario': scenario['scenario'],
                'response': data
            }
        
        except requests.exceptions.RequestException as e:
            error_msg = str(e)
            if hasattr(e, 'response') and e.response is not None:
                try:
                    error_data = e.response.json()
                    error_msg = error_data.get('error', str(e))
                except:
                    error_msg = e.response.text or str(e)
            
            print(f"❌ Error: {error_msg}")
            
            return {
                'success': False,
                'error': error_msg,
                'scenario': scenario['scenario']
            }
    
    def check_workflow_status(self, workflow_id: str) -> Dict[str, Any]:
        """Check the status of a workflow"""
        try:
            response = self.session.get(
                f"{self.api_url}/workflows/{workflow_id}/status",
                timeout=10
            )
            response.raise_for_status()
            return response.json()
        except Exception as e:
            print(f"Error checking status: {e}")
            return None
    
    def run_batch_tests(self, test_scenarios: List[Dict[str, Any]], 
                       test_name: str = "Test Scenarios") -> None:
        """Run a batch of test scenarios"""
        print(f"\n📋 Running {test_name}...")
        
        for scenario in test_scenarios:
            result = self.run_test(scenario)
            self.results['total'] += 1
            
            if result['success']:
                self.results['successful'].append(result)
            else:
                self.results['failed'].append(result)
            
            # Wait between tests
            time.sleep(DELAY_BETWEEN_TESTS)
    
    def generate_report(self) -> str:
        """Generate a detailed test report"""
        duration = datetime.now() - self.results['start_time']
        
        report = f"""
Security Testing Platform - Test Report
======================================
Generated: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}
Duration: {duration}

Summary
-------
Total Tests: {self.results['total']}
✅ Successful: {len(self.results['successful'])}
❌ Failed: {len(self.results['failed'])}
Success Rate: {(len(self.results['successful']) / self.results['total'] * 100):.1f}%
"""
        
        if self.results['failed']:
            report += "\n\nFailed Tests\n------------\n"
            for failure in self.results['failed']:
                report += f"- {failure['scenario']}\n"
                report += f"  Error: {failure['error']}\n\n"
        
        if self.results['successful']:
            report += "\n\nSuccessful Tests\n---------------\n"
            for success in self.results['successful']:
                report += f"- {success['scenario']}\n"
                report += f"  Workflow ID: {success['workflowId']}\n"
                
                # Check current status
                status = self.check_workflow_status(success['workflowId'])
                if status:
                    report += f"  Current Status: {status.get('status', 'Unknown')}\n"
                report += "\n"
        
        return report
    
    def save_results(self, filename: str = None) -> str:
        """Save test results to a file"""
        if filename is None:
            filename = f"test-results-{datetime.now().strftime('%Y%m%d-%H%M%S')}.json"
        
        with open(filename, 'w') as f:
            json.dump({
                'timestamp': datetime.now().isoformat(),
                'api_url': self.api_url,
                'results': self.results,
                'summary': {
                    'total': self.results['total'],
                    'successful': len(self.results['successful']),
                    'failed': len(self.results['failed'])
                }
            }, f, indent=2, default=str)
        
        print(f"\n💾 Results saved to: {filename}")
        return filename


def load_test_inputs(filename: str = 'test-user-inputs.json') -> Dict[str, Any]:
    """Load test inputs from JSON file"""
    with open(filename, 'r') as f:
        return json.load(f)


def main():
    parser = argparse.ArgumentParser(
        description='Run security platform tests'
    )
    parser.add_argument(
        '--scenario', 
        help='Run a specific scenario by name'
    )
    parser.add_argument(
        '--list', 
        action='store_true',
        help='List all available scenarios'
    )
    parser.add_argument(
        '--api-url',
        default=API_BASE_URL,
        help='API base URL'
    )
    parser.add_argument(
        '--save-results',
        action='store_true',
        help='Save test results to file'
    )
    parser.add_argument(
        '--report',
        action='store_true',
        help='Generate detailed report'
    )
    parser.add_argument(
        '--quick',
        action='store_true',
        help='Run only quick scan tests'
    )
    parser.add_argument(
        '--edge-cases-only',
        action='store_true',
        help='Run only edge case tests'
    )
    
    args = parser.parse_args()
    
    # Load test inputs
    test_inputs = load_test_inputs()
    
    if args.list:
        print("Available test scenarios:\n")
        print("Regular Scenarios:")
        for scenario in test_inputs['test_scenarios']:
            print(f"  - {scenario['scenario']}: {scenario['description']}")
        print("\nEdge Cases:")
        for scenario in test_inputs['edge_cases']:
            print(f"  - {scenario['scenario']}: {scenario['description']}")
        return
    
    # Initialize test runner
    runner = SecurityTestRunner(args.api_url)
    
    if args.scenario:
        # Run specific scenario
        all_scenarios = test_inputs['test_scenarios'] + test_inputs['edge_cases']
        scenario = next((s for s in all_scenarios if s['scenario'] == args.scenario), None)
        
        if not scenario:
            print(f"❌ Scenario '{args.scenario}' not found")
            print("\nUse --list to see available scenarios")
            return
        
        runner.run_test(scenario)
    
    else:
        # Run batch tests
        if args.edge_cases_only:
            runner.run_batch_tests(test_inputs['edge_cases'], "Edge Cases")
        elif args.quick:
            quick_tests = [s for s in test_inputs['test_scenarios'] 
                          if s['input']['testType'] == 'quick']
            runner.run_batch_tests(quick_tests, "Quick Scan Tests")
        else:
            runner.run_batch_tests(test_inputs['test_scenarios'], "Test Scenarios")
            runner.run_batch_tests(test_inputs['edge_cases'], "Edge Cases")
    
    # Generate report if requested
    if args.report:
        report = runner.generate_report()
        print(report)
        
        # Save report to file
        report_file = f"test-report-{datetime.now().strftime('%Y%m%d-%H%M%S')}.txt"
        with open(report_file, 'w') as f:
            f.write(report)
        print(f"\n📄 Report saved to: {report_file}")
    
    # Save results if requested
    if args.save_results:
        runner.save_results()
    
    # Print summary
    print(f"\n\n📊 Test Summary")
    print("===============")
    print(f"Total Tests: {runner.results['total']}")
    print(f"✅ Successful: {len(runner.results['successful'])}")
    print(f"❌ Failed: {len(runner.results['failed'])}")


if __name__ == '__main__':
    main()

File: test-user-input-runner.js
#!/usr/bin/env node

/**
 * Test User Input Runner
 * This script demonstrates how to programmatically submit test cases to the security testing platform
 */

import axios from 'axios';
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const testInputs = JSON.parse(readFileSync(join(__dirname, 'test-user-inputs.json'), 'utf-8'));

// Configuration
const API_BASE_URL = process.env.API_URL || 'http://localhost:3000/api';
const DELAY_BETWEEN_TESTS = 2000; // 2 seconds between tests

// Helper function to create the API payload
function createWorkflowPayload(testInput) {
  const payload = {
    target: testInput.target,
    scope: testInput.scope,
    description: testInput.description,
    template: `security-${testInput.testType}`
  };

  // Add authentication if provided
  if (testInput.username && testInput.password) {
    payload.auth = {
      username: testInput.username,
      password: testInput.password
    };
  }

  return payload;
}

// Function to run a single test
async function runTest(scenario) {
  console.log(`\n🔍 Running: ${scenario.scenario}`);
  console.log(`📝 Description: ${scenario.description}`);
  
  try {
    const payload = createWorkflowPayload(scenario.input);
    console.log('📤 Payload:', JSON.stringify(payload, null, 2));
    
    const response = await axios.post(`${API_BASE_URL}/run-soc2-workflow`, payload);
    
    console.log('✅ Success! Workflow ID:', response.data.workflowId);
    console.log('📊 Status:', response.data.status);
    
    return {
      success: true,
      workflowId: response.data.workflowId,
      scenario: scenario.scenario
    };
  } catch (error) {
    console.error('❌ Error:', error.response?.data?.error || error.message);
    
    return {
      success: false,
      error: error.response?.data?.error || error.message,
      scenario: scenario.scenario
    };
  }
}

// Function to check workflow status
async function checkWorkflowStatus(workflowId) {
  try {
    const response = await axios.get(`${API_BASE_URL}/workflows/${workflowId}/status`);
    return response.data;
  } catch (error) {
    console.error('Error checking status:', error.message);
    return null;
  }
}

// Main execution function
async function runAllTests() {
  console.log('🚀 Starting Security Platform Test Suite');
  console.log('=====================================\n');
  
  const results = {
    successful: [],
    failed: [],
    total: 0
  };

  // Run regular test scenarios
  console.log('📋 Running Test Scenarios...');
  for (const scenario of testInputs.test_scenarios) {
    const result = await runTest(scenario);
    results.total++;
    
    if (result.success) {
      results.successful.push(result);
    } else {
      results.failed.push(result);
    }
    
    // Wait between tests to avoid overwhelming the system
    await new Promise(resolve => setTimeout(resolve, DELAY_BETWEEN_TESTS));
  }

  // Run edge cases
  console.log('\n\n📋 Running Edge Cases...');
  for (const edgeCase of testInputs.edge_cases) {
    const result = await runTest(edgeCase);
    results.total++;
    
    if (result.success) {
      results.successful.push(result);
    } else {
      results.failed.push(result);
    }
    
    await new Promise(resolve => setTimeout(resolve, DELAY_BETWEEN_TESTS));
  }

  // Print summary
  console.log('\n\n📊 Test Summary');
  console.log('===============');
  console.log(`Total Tests: ${results.total}`);
  console.log(`✅ Successful: ${results.successful.length}`);
  console.log(`❌ Failed: ${results.failed.length}`);
  
  if (results.failed.length > 0) {
    console.log('\n❌ Failed Scenarios:');
    results.failed.forEach(failure => {
      console.log(`  - ${failure.scenario}: ${failure.error}`);
    });
  }
  
  if (results.successful.length > 0) {
    console.log('\n✅ Successful Workflows:');
    for (const success of results.successful) {
      console.log(`  - ${success.scenario}: ${success.workflowId}`);
      
      // Check status of successful workflows
      const status = await checkWorkflowStatus(success.workflowId);
      if (status) {
        console.log(`    Status: ${status.status}`);
      }
    }
  }
}

// Run a specific scenario by name
async function runSpecificScenario(scenarioName) {
  const allScenarios = [...testInputs.test_scenarios, ...testInputs.edge_cases];
  const scenario = allScenarios.find(s => s.scenario === scenarioName);
  
  if (!scenario) {
    console.error(`❌ Scenario "${scenarioName}" not found`);
    console.log('\nAvailable scenarios:');
    allScenarios.forEach(s => console.log(`  - ${s.scenario}`));
    return;
  }
  
  await runTest(scenario);
}

// Command line interface
const args = process.argv.slice(2);

if (args.length === 0) {
  // Run all tests
  runAllTests().catch(console.error);
} else if (args[0] === '--scenario' && args[1]) {
  // Run specific scenario
  runSpecificScenario(args[1]).catch(console.error);
} else if (args[0] === '--list') {
  // List all scenarios
  console.log('Available test scenarios:');
  console.log('\nRegular Scenarios:');
  testInputs.test_scenarios.forEach(s => {
    console.log(`  - ${s.scenario}: ${s.description}`);
  });
  console.log('\nEdge Cases:');
  testInputs.edge_cases.forEach(s => {
    console.log(`  - ${s.scenario}: ${s.description}`);
  });
} else {
  console.log('Usage:');
  console.log('  node test-user-input-runner.js              # Run all tests');
  console.log('  node test-user-input-runner.js --list       # List all scenarios');
  console.log('  node test-user-input-runner.js --scenario "Scenario Name"  # Run specific scenario');
}

export {
  createWorkflowPayload,
  runTest,
  checkWorkflowStatus
};